### PR TITLE
Enforce element types of DASH containers to be trivially copyable

### LINKED ISSUE #241: dash::Matrix doesn't accept struct as template datatype
Why does `dash::Matrix` and thus `dash::NArray` not accept struct as template datatype while `dash::Array` does?
For example:

~~~c++
using Point = struct{ MTRX_TYPE value;
                      uint row, col;};
dash::Array  < Point   > points( nelts );  // ok
dash::NArray < Point,1 > points( nelts );  // bad
dash::Matrix < Point,1 > points( nelts );  // bad
~~~

### PR DESCRIPTION
**Pull request by devreal**
Consistently enforce `std::is_trivially_copyable` on element types of containers (except when using the Cray compiler).
Introduce minimum required compiler version.
Makes `GlobPtr` trivially copyable after #221 has been resolved.
Introduce `dash::Pair` which closely resembles `std::pair` except for being trivially copyable. 
Fixes #241

### CODE REVIEW

**Comment by fuchsto** [comment]
My Clangs also have issues with `is_trivially_copyable`. Perhaps we should introduce our own type traits for portability.

File: dash/include/dash/Matrix.h [hunk 1]
```diff
@@ -135,13 +135,16 @@ template<
   class    PatternT       = TilePattern<NumDimensions, ROW_MAJOR, IndexT> >
 class Matrix
 {
-#if !defined(_CRAYC) && !defined(__INTEL_COMPILER)
```
Comment (fuchsto) [inline dash/include/dash/Matrix.h:138]: Intel compilers (tested 16.0 and 17.0) don't allow `std::is_trivially_copyable` either. Why did you remove this check?

**Review by fuchsto — CHANGES_REQUESTED** [review]

File: dash/include/dash/Matrix.h (hunk 1 above)
Comment (devreal) [reply dash/include/dash/Matrix.h:138]: Because they do. I just tested with Intel 17.0 and I am sure I also verified 16 back when we discussed #241. Keep in mind that the Intel compiler relies on the GNU-provided STL so you might have to load a newer GCC module (at least I have to do that on our Linux cluster).

**Comment by devreal** [comment]
@fuchsto Just checked with Clang 3.8.0, works fine for me. What version did you test with? 

How would you check trivial-copyable in a custom type trait? `std::is_trivially_copyable` relies on the compiler builtin function `__is_trivially_copyable`...

**Comment by fuchsto** [comment]
I don't see what you mean.
Type traits are not checked, they are defined. So, we would define additional traits in DASH, like we did with `is_atomic`.

**Comment by fuchsto** [comment]
@devreal We could define traits depending on the compiler and its version. If this is just an issue with Cray compilers, it's not worth the effort, of course.

File: dash/include/dash/Matrix.h (hunk 1 above)
Comment (fuchsto) [reply dash/include/dash/Matrix.h:138]: Yes, that seems alright.
I actually did test exactly that, but it's apparently specific to our dev server. When reloading gcc after loading the icc module, there is some hickup and the newer version of gcc is not used.

**Review by fuchsto — APPROVED** [review]
